### DISCUSSION TITLE
How to initiate an FSM outside a message handler?

### DISCUSSION THREAD

**Discussion by nanodesuologist** [aiogram/aiogram#530] [Q&A]
How do I initiate an FSM outside a message handler? E.g. in on_startup().

I would like to set the state for a user on startup but the error is:
```ValueError: `user` or `chat` parameter is required but no one is provided!```
And I can't even specify what user I want to set the state for.
```python
from aiogram import Bot, Dispatcher, executor
from aiogram.contrib.fsm_storage.memory import MemoryStorage
from aiogram.dispatcher.filters.state import State, StatesGroup

dp = Dispatcher(Bot(token=...), storage=MemoryStorage())


class Test(StatesGroup):
    step_1 = State()
    step_2 = State()


@dp.message_handler(state=Test.step_1)
async def process_step_1(message, state):
    ...


def on_startup(dp):
    chat = ...
    await dp.bot.send_message(chat, 'hi')
    await Test.first()  # Error


executor.start_polling(dp, on_startup=on_startup)
```

**Comment by darksidecat** [comment]
You can work directly with storage and set state for chats/users:

https://github.com/aiogram/aiogram/blob/cce29ba532b4956557dead60edf14e5d0e5340e0/aiogram/contrib/fsm_storage/memory.py#L56-L61

example:

```python
import asyncio

from aiogram import Bot, Dispatcher, types
from aiogram.contrib.fsm_storage.memory import MemoryStorage
from aiogram.dispatcher import FSMContext
from aiogram.dispatcher.filters.state import StatesGroup, State


class Test(StatesGroup):
    test = State()


async def on_startup(storage):
    id_example = 999999999
    await storage.set_state(user=id_example , chat=id_example , state=Test.test.state)


async def main():
    bot = Bot("TOKEN")
    storage = MemoryStorage()
    dp = Dispatcher(bot, storage=storage)

    await on_startup(storage)

    @dp.message_handler(state="*")
    async def echo(msg: types.Message, state: FSMContext):
        print(await state.get_state())
        await msg.answer(f"Your state: {await state.get_state()}")

    try:
        await dp.start_polling()
    finally:
        await bot.session.close()


if __name__ == '__main__':
    try:
        asyncio.run(main())
    except (KeyboardInterrupt, SystemExit):
        pass

```

Output:
```
Test:test
```

**Reply by nanodesuologist** [reply]
What if I need to trigger a state handler without user's interaction whatsoever?

**Reply by darksidecat** [reply]
I'm not sure I understood the question. but in the example that I sent the echo-handler is only needed to display the current state. The state itself is set in on_startup

**Reply by nanodesuologist** [reply]
Yeah, but if you wanted to add a handler to that state, e.g.
```python
@dp.message_handler(state=Test.test):
async def test(message):
    await message.answer('test')
```
Then it would not be executed after you set it that way

**Reply by darksidecat** [reply]
> Then it would not be executed after you set it that way

Just tested that with handler like this, state was setted and handler work just like planned
```python
@dp.message_handler(state=Test.test)
    async def echo(msg: types.Message, state: FSMContext):
        await msg.answer(f"Your state: {await state.get_state()}")
```

Did you specify the chat+user ID exactly when setting the state in on_startup?
The state for the user/chat can only be set when these ids are passed.

But if you need the default value of the state for all users (for new ones or if you do not have the data id), then it already exists: None

Also update aiogram to the latest version just in case.

**Reply by nanodesuologist** [reply]
I have tested this exact example, the handler didn't run for me, aiogram==2.12.1
